### PR TITLE
Fix assignment of type for ConsumerToken subclasses

### PR DESCRIPTION
**Pull request by afeld**
ActiveRecord (at least in Rails 3) doesn't allow you to set/query records on an explicit value for the `type` field.

### CODE REVIEW

**Comment by juni0r** [comment]
Yep. I can confirm this.

A quick fix is to override the method in the generated ConsumerToken class:

```
class ConsumerToken < ActiveRecord::Base

  ...

  def self.find_or_create_from_access_token(user,access_token)
    if user
      token = self.find_or_create_by_user_id_and_token(user.id, access_token.token)
    else
      token = self.find_or_create_by_token(access_token.token)
    end

    if token.new_record?
      token.secret = access_token.secret
      token.save
    end

    token
  end
end
```

**Comment by flippyhead** [comment]
Yes, this is a problem thanks

**Comment by pelle** [comment]
Sorry. I had already fixed this. Should have looked through the pull requests first. Please test that master works for you .

File: lib/oauth/models/consumers/token.rb [hunk 1]
```diff
@@ -40,12 +40,17 @@ module Oauth
...
-              user.consumer_tokens.first(:conditions=>{:type=>self.to_s,:token=>access_token.token}) ||
-                user.consumer_tokens.create!(:type=>self.to_s,:token=>access_token.token, :secret=>access_token.secret)
+              token = self.find_or_create_by_user_id_and_token(user.id, access_token.token)
             else
-              ConsumerToken.first( :conditions =>{ :token=>access_token.token,:type=>self.to_s}) ||
-                create(:type=>self.to_s,:token=>access_token.token, :secret=>access_token.secret)
+              token = self.find_or_create_by_token(access_token.token)
             end
+            
+            if token.new_record?
```
Comment (afeld) [inline lib/oauth/models/consumers/token.rb]: this would never evaluate to true

File: lib/oauth/models/consumers/token.rb [hunk 2]
```diff
@@ -40,12 +40,16 @@ module Oauth
...
+              token = self.find_or_initialize_by_user_id_and_token(user.id, access_token.token)
             else
-              ConsumerToken.first( :conditions =>{ :token=>access_token.token,:type=>self.to_s}) ||
-                create(:type=>self.to_s,:token=>access_token.token, :secret=>access_token.secret)
+              token = self.find_or_initialize_by_token(access_token.token)
             end
+            
+            # set or update the secret
+            token.secret = access_token.secret
+            token.save!
```
Comment (afeld) [inline lib/oauth/models/consumers/token.rb:50]: really, we want

```
token.save! if token.new_record? or token.changed?
```

is there a shorthand in activerecord for this?  maybe `update_attribute`?

**Comment by pelle** [comment]
I've included this now. Please test it and see if it works.

**Comment by afeld** [comment]
looks good, thanks!

**Pull request closed without merging**